cmd/src/campaign_progress_printer_test.go

```diff
@@ -0,0 +1,249 @@
+	if runtime.GOOS == "windows" {
+		t.Skip("Something emits different escape codes on windows.")
+	}
+
+	buf := &ttyBuf{}
+
+	out := output.NewOutput(buf, output.OutputOpts{
+		ForceTTY:    true,
+		ForceColor:  true,
+		ForceHeight: 25,
+		ForceWidth:  80,
+		Verbose:     true,
+	})
+
+	statuses := []*campaigns.TaskStatus{
+		{
+			RepoName:           "github.com/sourcegraph/sourcegraph",
+			StartedAt:          time.Now(),
+			CurrentlyExecuting: "echo Hello World > README.md",
+		},
+		{
+			RepoName:           "github.com/sourcegraph/src-cli",
+			StartedAt:          time.Now().Add(time.Duration(-5) * time.Second),
+			CurrentlyExecuting: "Downloading archive",
+		},
+		{
+			RepoName:           "github.com/sourcegraph/automation-testing",
+			StartedAt:          time.Now().Add(time.Duration(-5) * time.Second),
+			CurrentlyExecuting: "echo Hello World > README.md",
+		},
+	}
+
+	printer := newCampaignProgressPrinter(out, true, 4)
+	printer.forceNoSpinner = true
+
+	// Print with all three tasks running
+	printer.PrintStatuses(statuses)
+	have := buf.Lines()
+	want := []string{
+		"⠋  Executing... (0/3, 0 errored)  ",
+		"│                                                                               ",
+		"├── github.com/sourcegraph/sourcegraph  echo Hello World > README.md          0s",
+		"├── github.com/sourcegraph/src-cli      Downloading archive                   0s",
+		"└── github.com/sourcegraph/automati...  echo Hello World > README.md          0s",
+	}
+	if !cmp.Equal(want, have) {
+		t.Fatalf("wrong output:\n%s", cmp.Diff(want, have))
+	}
+
+	// Now mark the last task as completed
+	statuses[len(statuses)-1] = &campaigns.TaskStatus{
+		RepoName:           "github.com/sourcegraph/automation-testing",
+		StartedAt:          time.Now().Add(time.Duration(-5) * time.Second),
+		FinishedAt:         time.Now().Add(time.Duration(5) * time.Second),
+		CurrentlyExecuting: "",
+		Err:                nil,
+		ChangesetSpec: &campaigns.ChangesetSpec{
+			BaseRepository: "graphql-id",
+			CreatedChangeset: &campaigns.CreatedChangeset{
+				BaseRef:        "refs/heads/main",
+				BaseRev:        "d34db33f",
+				HeadRepository: "graphql-id",
+				HeadRef:        "refs/heads/my-campaign",
+				Title:          "This is my campaign",
+				Body:           "This is my campaign",
+				Commits: []campaigns.GitCommitDescription{
+					{
+						Message: "This is my campaign",
+						Diff:    progressPrinterDiff,
+					},
+				},
+				Published: false,
+			},
+		},
+	}
+
+	printer.PrintStatuses(statuses)
+	have = buf.Lines()
+	want = []string{
+		"github.com/sourcegraph/automation-testing",
+		"\tREADME.md   | 3 +++",
+		"\ta/b/c/c.txt | 1 -",
+		"\tx/x.txt     | 2 +-",
+		"  3 files changed, 4 insertions, 2 deletions",
+		"  Execution took 10s",
+		"",
+		"⠋  Executing... (1/3, 0 errored)  ███████████████▍",
+		"│                                                                                 0s", // Not sure why this is here, bug?
```

Comment:
Did you ever figure out what this was there? (Just so I'm clear: the bug is the extra `0s` that doesn't appear to correspond to any task?)

Reply:
Turns out it was a bug in our code! See: 550dafa I never cleared the current line when printing the `|        ` blank line. And the `ttyBuf` implementation keeps all previously written things in memory, so this was a leftover. The reason why this wasn't visible in the actual TUI I'm not sure, but I suspect it's because the `ttyBuf` handles lines differently

Reply:
Cool, thanks!